src/revlanguage/workspace/RevVariable.cpp

```diff
@@ -154,16 +121,19 @@ RevVariable& RevVariable::operator=(const RevVariable &v)
 
 
 /** Resize the vector to include this index. */
-void RevVariable::addIndex(size_t idx)
+void RevVariable::addIndex(int idx, const RevPtr<RevVariable>& element)
 {
-    
-    if ( idx > element_index_max )
-    {
-        element_index_max = idx;
-    }
+    assert( isVectorVariable() );
```

Comment:
Any reason to prefer `assert` here over `RbException`? `RbException` might be a more pleasant user experience (less unpleasant anyway), given the context here is user-defined functions!

Reply:
I'm thinking of this as documenting a precondition, which is a common use case for asserts.  

Usually the standard is to use asserts for programmer errors and report run-time error messages for user errors.  Any code that calls `addIndex( )` without first checking that it is operating on a vector indicates a programmer error, not a user error.  Probably the existing code has already checked whether the variable is a vector.

RevBayes' lack of asserts indicates low engineering standards, according to the Stamatakis group's paper "The State of Software for Evolutionary Biology".  See Tables 2 and 3 where they score software according to the number of asserts per 1000 lines of code.  I think most programmers would agree with that paper.

Kids these day would call the lack of asserts a "code smell" :roll_eyes: 

Thoughts?

Reply:
That's fair enough, but you could make the same argument that attempting to assign an idx < 1 is a programming error? So maybe that should be an assert, too?

Reply:
I see what you mean.

I'm not sure if `idx < 1` is actually a programmer error the way RevBayes is currently written :grimacing:  
In some cases, an `idx` of `0` can be passed down to `addIndex( )`, and in some other cases it is intercepted before it gets there.  I think its kind of a mess with no clear contract to enforce by assert.

Perhaps figuring out how indices are passed down to `addIndex( )` would be a good thing to unsnarl later, but I didn't want to make this patch bigger than it is.

Reply:
Short version: that's a good question, but I think that `(idx < 1)` is a user error and not a programmer error.  Thoughts?

Reply:
Yes I see. I agree with this distinction now.